File: server/sdk/v1/python/proto_fleet_sdk/protocols/device.py
"""Device protocol interfaces.

This module defines all device-level protocols that device implementations should satisfy.
The Device protocol is a composite of multiple sub-protocols for different functional areas.
"""

from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING, Protocol

import grpc

from proto_fleet_sdk.enums import CoolingMode, PerformanceMode
from proto_fleet_sdk.types import (
    Capabilities,
    ConfiguredPool,
    DeviceInfo,
    FirmwareFile,
    MiningPoolConfig,
)

if TYPE_CHECKING:
    from proto_fleet_sdk.error_codes import DeviceErrors
    from proto_fleet_sdk.telemetry.metrics import DeviceMetrics

__all__ = [
    "DeviceCore",
    "DeviceControl",
    "DeviceConfiguration",
    "DeviceMaintenance",
    "DeviceErrorReporting",
    "Device",
]


class DeviceCore(Protocol):
    """Core device interface methods required for all devices."""

    def id(self) -> str:
        ...

    async def describe_device(self, ctx: grpc.ServicerContext) -> tuple[DeviceInfo, Capabilities]:
        ...

    async def status(self, ctx: grpc.ServicerContext) -> DeviceMetrics:
        """Get current device telemetry.

        Returns a complete DeviceMetrics snapshot with health status, aggregated
        metrics, and component-level telemetry.
        """
        ...

    async def close(self, ctx: grpc.ServicerContext) -> None:
        """Close device and release resources.

        Called when the device is being removed from the fleet or the plugin is
        shutting down. Should close connections, cancel background tasks, and
        clean up any other resources.
        """
        ...


class DeviceControl(Protocol):
    """Device control operations for mining and diagnostics."""

    async def start_mining(self, ctx: grpc.ServicerContext) -> None:
        ...

    async def stop_mining(self, ctx: grpc.ServicerContext) -> None:
        ...

    async def blink_led(self, ctx: grpc.ServicerContext) -> None:
        ...

    async def reboot(self, ctx: grpc.ServicerContext) -> None:
        ...


class DeviceConfiguration(Protocol):
    """Device configuration operations."""

    async def set_cooling_mode(self, ctx: grpc.ServicerContext, mode: CoolingMode) -> None:
        ...

    async def get_cooling_mode(self, ctx: grpc.ServicerContext) -> CoolingMode:
        ...

    async def set_power_target(self, ctx: grpc.ServicerContext, performance_mode: PerformanceMode) -> None:
        ...

    async def update_mining_pools(self, ctx: grpc.ServicerContext, pools: list[MiningPoolConfig]) -> None:
        ...

    async def get_mining_pools(self, ctx: grpc.ServicerContext) -> list[ConfiguredPool]:
        ...

    async def update_miner_password(
        self, ctx: grpc.ServicerContext, current_password: str, new_password: str
    ) -> None:
        ...


class DeviceMaintenance(Protocol):
    """Device maintenance operations."""

    async def download_logs(
        self, ctx: grpc.ServicerContext, since: datetime | None = None, batch_log_uuid: str | None = None
    ) -> tuple[str, bool]:
        """Download device logs.

        Returns (log_data, more_data) where more_data indicates if additional
        logs are available with the same batch_log_uuid.
        """
        ...

    async def firmware_update(self, ctx: grpc.ServicerContext, firmware: FirmwareFile) -> None:
        ...

    async def unpair(self, ctx: grpc.ServicerContext) -> None:
        """Unpair device and clear credentials.

        Clears any stored credentials or configuration related to the
        Fleet management system from the device.
        """
        ...


class DeviceErrorReporting(Protocol):
    """Device error reporting interface."""

    async def get_errors(self, ctx: grpc.ServicerContext) -> DeviceErrors:
        ...


class Device(
    DeviceCore,
    DeviceControl,
    DeviceConfiguration,
    DeviceMaintenance,
    DeviceErrorReporting,
    Protocol,
):
    """Complete device interface composing all sub-protocols.

    A device implementation should satisfy all these protocols. Methods that are not
    supported by the hardware should raise UnsupportedCapabilityError.
    """

    pass

File: plugin/virtual/internal/device/device.go
// Package device implements the SDK Device interface for virtual miners.
package device

import (
	"context"
	"log/slog"
	"sync"
	"time"

	"github.com/block/proto-fleet/plugin/virtual/internal/config"
	"github.com/block/proto-fleet/plugin/virtual/pkg/virtual"
	sdk "github.com/block/proto-fleet/server/sdk/v1"
)

// Compile-time check that *Device implements sdk.Device interface.
var _ sdk.Device = (*Device)(nil)

const statusCacheTTL = 5 * time.Second

// Device implements sdk.Device for a virtual miner.
type Device struct {
	id         string
	deviceInfo sdk.DeviceInfo
	config     *config.VirtualMinerConfig
	simulator  *virtual.Simulator

	// Simulated state
	isMining        bool
	coolingMode     sdk.CoolingMode
	performanceMode sdk.PerformanceMode
	pools           []sdk.MiningPoolConfig

	// curtailLevel is nonzero while telemetry should report inactive mining.
	curtailLevel           sdk.CurtailLevel
	preCurtailMiningActive *bool

	// Status caching
	lastStatus   *sdk.DeviceMetrics
	lastStatusAt time.Time

	mutex sync.Mutex
}

// New creates a new virtual device instance.
func New(id string, deviceInfo sdk.DeviceInfo, cfg *config.VirtualMinerConfig) *Device {
	return &Device{
		id:              id,
		deviceInfo:      deviceInfo,
		config:          cfg,
		simulator:       virtual.NewSimulator(cfg),
		isMining:        true, // Start in mining state by default
		coolingMode:     sdk.CoolingModeAirCooled,
		performanceMode: sdk.PerformanceModeMaximumHashrate,
		pools:           []sdk.MiningPoolConfig{},
	}
}

// ID implements sdk.DeviceCore.
func (d *Device) ID() string {
	return d.id
}

// DescribeDevice implements sdk.DeviceCore.
func (d *Device) DescribeDevice(_ context.Context) (sdk.DeviceInfo, sdk.Capabilities, error) {
	return d.deviceInfo, sdk.Capabilities{
		sdk.CapabilityPollingHost:       true,
		sdk.CapabilityReboot:            true,
		sdk.CapabilityMiningStart:       true,
		sdk.CapabilityMiningStop:        true,
		sdk.CapabilityLEDBlink:          true,
		sdk.CapabilityCoolingModeAir:    true,
		sdk.CapabilityPoolConfig:        true,
		sdk.CapabilityHashrateReported:  true,
		sdk.CapabilityPowerUsage:        true,
		sdk.CapabilityTemperature:       true,
		sdk.CapabilityFanSpeed:          true,
		sdk.CapabilityEfficiency:        true,
		sdk.CapabilityPerBoardStats:     true,
		sdk.CapabilityPSUStats:          true,
		sdk.CapabilityRealtimeTelemetry: true,
		// v1 advertises FULL curtailment only.
		sdk.CapabilityCurtailFull: true,
	}, nil
}

// Status implements sdk.DeviceCore.
func (d *Device) Status(_ context.Context) (sdk.DeviceMetrics, error) {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	// Return cached status if still valid
	if d.lastStatus != nil && time.Since(d.lastStatusAt) < statusCacheTTL {
		return *d.lastStatus, nil
	}

	// Generate new metrics
	metrics := d.simulator.GenerateMetrics(d.id, d.isMining)
	d.lastStatus = &metrics
	d.lastStatusAt = time.Now()

	return metrics, nil
}

// Close implements sdk.DeviceCore.
func (d *Device) Close(_ context.Context) error {
	slog.Info("Closing virtual device", "device_id", d.id)
	return nil
}

// StartMining implements sdk.DeviceControl.
func (d *Device) StartMining(_ context.Context) error {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	d.isMining = true
	d.clearCurtailmentStateLocked()
	d.lastStatus = nil // Invalidate cache
	slog.Info("Virtual miner started mining", "device_id", d.id)
	return nil
}

// StopMining implements sdk.DeviceControl.
func (d *Device) StopMining(_ context.Context) error {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	d.isMining = false
	d.clearCurtailmentStateLocked()
	d.lastStatus = nil
	slog.Info("Virtual miner stopped mining", "device_id", d.id)
	return nil
}

// BlinkLED implements sdk.DeviceControl.
func (d *Device) BlinkLED(_ context.Context) error {
	slog.Info("Virtual miner LED blink triggered", "device_id", d.id)
	return nil
}

// Reboot implements sdk.DeviceControl.
func (d *Device) Reboot(_ context.Context) error {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	slog.Info("Virtual miner rebooting", "device_id", d.id)

	// Simulate brief downtime
	d.isMining = false
	d.lastStatus = nil

	// Immediately come back up (in a real scenario you might add a delay)
	d.isMining = true

	return nil
}

// Curtail honors FULL and rejects reserved levels.
func (d *Device) Curtail(_ context.Context, req sdk.CurtailRequest) error {
	if req.Level != sdk.CurtailLevelFull {
		return sdk.NewErrCurtailCapabilityNotSupported(d.id, int32(req.Level))
	}

	d.mutex.Lock()
	defer d.mutex.Unlock()

	if d.preCurtailMiningActive == nil {
		wasMining := d.isMining
		d.preCurtailMiningActive = &wasMining
	}
	d.curtailLevel = req.Level
	d.isMining = false
	d.lastStatus = nil
	slog.Info("Virtual miner curtailed", "device_id", d.id, "level", req.Level)
	return nil
}

// Uncurtail clears curtailment; duplicate calls are no-ops.
func (d *Device) Uncurtail(_ context.Context, _ sdk.UncurtailRequest) error {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	if d.curtailLevel == sdk.CurtailLevelUnspecified {
		slog.Info("Virtual miner uncurtail requested while not curtailed (no-op)", "device_id", d.id)
		return nil
	}
	shouldMine := true
	if d.preCurtailMiningActive != nil {
		shouldMine = *d.preCurtailMiningActive
	}
	d.curtailLevel = sdk.CurtailLevelUnspecified
	d.preCurtailMiningActive = nil
	d.isMining = shouldMine
	d.lastStatus = nil
	slog.Info("Virtual miner uncurtailed", "device_id", d.id)
	return nil
}

func (d *Device) clearCurtailmentStateLocked() {
	d.curtailLevel = sdk.CurtailLevelUnspecified
	d.preCurtailMiningActive = nil
}

// GetCoolingMode implements sdk.DeviceConfiguration.
func (d *Device) GetCoolingMode(_ context.Context) (sdk.CoolingMode, error) {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	return d.coolingMode, nil
}

// SetCoolingMode implements sdk.DeviceConfiguration.
func (d *Device) SetCoolingMode(_ context.Context, mode sdk.CoolingMode) error {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	d.coolingMode = mode
	slog.Info("Virtual miner cooling mode set", "device_id", d.id, "mode", mode)
	return nil
}

// SetPowerTarget implements sdk.DeviceConfiguration.
func (d *Device) SetPowerTarget(_ context.Context, mode sdk.PerformanceMode) error {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	d.performanceMode = mode
	slog.Info("Virtual miner performance mode set", "device_id", d.id, "mode", mode)
	return nil
}

// UpdateMiningPools implements sdk.DeviceConfiguration.
func (d *Device) UpdateMiningPools(_ context.Context, pools []sdk.MiningPoolConfig) error {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	d.pools = pools
	slog.Info("Virtual miner pools updated", "device_id", d.id, "pool_count", len(pools))
	return nil
}

// GetMiningPools implements sdk.DeviceConfiguration.
func (d *Device) GetMiningPools(_ context.Context) ([]sdk.ConfiguredPool, error) {
	d.mutex.Lock()
	defer d.mutex.Unlock()

	result := make([]sdk.ConfiguredPool, len(d.pools))
	for i, p := range d.pools {
		result[i] = sdk.ConfiguredPool{
			Priority: p.Priority,
			URL:      p.URL,
			Username: p.WorkerName,
		}
	}
	return result, nil
}

// UpdateMinerPassword implements sdk.DeviceConfiguration.
func (d *Device) UpdateMinerPassword(_ context.Context, _, _ string) error {
	slog.Info("Virtual miner password update requested (no-op)", "device_id", d.id)
	return nil
}

// DownloadLogs implements sdk.DeviceMaintenance.
func (d *Device) DownloadLogs(_ context.Context, _ *time.Time, _ string) (string, bool, error) {
	return "Virtual miner log data - no actual logs available", false, nil
}

// FirmwareUpdate implements sdk.DeviceMaintenance.
func (d *Device) FirmwareUpdate(_ context.Context, _ sdk.FirmwareFile) error {
	slog.Info("Virtual miner firmware update requested (no-op)", "device_id", d.id)
	return nil
}

// Unpair implements sdk.DeviceMaintenance.
func (d *Device) Unpair(_ context.Context) error {
	slog.Info("Virtual miner unpaired", "device_id", d.id)
	return nil
}

// GetErrors implements sdk.DeviceErrorReporting.
// Virtual miners report errors based on error injection configuration.
// Currently returns an empty list - error injection affects telemetry health status instead.
func (d *Device) GetErrors(_ context.Context) (sdk.DeviceErrors, error) {
	return sdk.DeviceErrors{
		DeviceID: d.id,
		Errors:   []sdk.DeviceError{},
	}, nil
}

// TryBatchStatus implements sdk.DeviceOptional.
func (d *Device) TryBatchStatus(_ context.Context, _ []string) (map[string]sdk.DeviceMetrics, bool, error) {
	return nil, false, nil
}

// TrySubscribe implements sdk.DeviceOptional.
func (d *Device) TrySubscribe(_ context.Context, _ []string) (<-chan sdk.DeviceMetrics, bool, error) {
	return nil, false, nil
}

// TryGetWebViewURL implements sdk.DeviceOptional.
func (d *Device) TryGetWebViewURL(_ context.Context) (string, bool, error) {
	return "", false, nil
}

// TryGetTimeSeriesData implements sdk.DeviceOptional.
func (d *Device) TryGetTimeSeriesData(_ context.Context, _ []string, _, _ time.Time, _ *time.Duration, _ int32, _ string) ([]sdk.DeviceMetrics, string, bool, error) {
	return nil, "", false, nil
}
